### PR TITLE
Fix failure in L1Trigger/TrackFindingTracklet reported in ASAN

### PR DESCRIPTION
**Pull request by mrodozov**
#### PR description:

Fixes at least one unit test in ASAN IB 
SLHCUpgradeSimulations/Geometry -> testPhase2PixelNtuple
Using what #30913 suggests

#### PR validation:

the unit test runs

### CODE REVIEW

**Comment by mmusich** [comment]
Duplicate of https://github.com/cms-sw/cmssw/pull/31554 ?
In any case the PR title is misleading as the package involved is actually L1Trigger/TrackFindingTracklet and the unit test just happens to consume it.

**Comment by mrodozov** [comment]
please test

**Comment by mrodozov** [comment]
please test

**Comment by mrodozov** [comment]
> Duplicate of #31554 ?
> In any case the PR title is misleading as the package involved is actually L1Trigger/TrackFindingTracklet and the unit test just happens to consume it.


yes it is but I've removed the redundancies and changed the misleading title.

**Comment by mrodozov** [comment]
abort
got stuck again

**Comment by mrodozov** [comment]
please test

**Comment by smuzaffar** [comment]
@mrodozov , I think due to this change it is going in to some infinite loop. Please run wf 23234.0 locally to see if it finishes

File: L1Trigger/TrackFindingTracklet/src/FitTrack.cc [hunk 1]
```diff
@@ -933,48 +933,40 @@ void FitTrack::execute() {
...
       nMatchesUniq++;
     match = false;
 
-    if (indexArray[2] < matches3.size()) {
-      while (matches3[indexArray[2]] == bestTracklet && indexArray[2] < matches3.size()) {
-        indexArray[2]++;
-        nMatches++;
-        match = true;
-      }
+    while (indexArray[2] < matches2.size() && matches2[indexArray[2]] == bestTracklet) {
```
Comment (smuzaffar) [inline L1Trigger/TrackFindingTracklet/src/FitTrack.cc:956]: @mrodozov , this is bug. Please use `matches3`

File: L1Trigger/TrackFindingTracklet/src/FitTrack.cc [hunk 2]
```diff
@@ -933,48 +933,40 @@ void FitTrack::execute() {
...
       nMatchesUniq++;
     match = false;
 
-    if (indexArray[3] < matches4.size()) {
-      while (matches4[indexArray[3]] == bestTracklet && indexArray[3] < matches4.size()) {
-        indexArray[3]++;
-        nMatches++;
-        match = true;
-      }
+    while (indexArray[3] < matches2.size() && matches2[indexArray[3]] == bestTracklet) {
```
Comment (smuzaffar) [inline L1Trigger/TrackFindingTracklet/src/FitTrack.cc:966]: same here, use `matches4`

**Comment by mrodozov** [comment]
abort

**Comment by mrodozov** [comment]
please test

**Comment by smuzaffar** [comment]
@rekovic, @benkrikler can you please review this?
@qliphy , @silviodonato this fixes most of ASAN IBs relvals failures. This change here does not change the ebhaviour of code. It just makes sure that we check for the size before accessing the memory.

**Comment by rekovic** [comment]
+1

**Comment by rekovic** [comment]
Attn @skinnari @tomalin

**Comment by silviodonato** [comment]
+1

**Pull request merged**